pkg/util/pki/kube_test.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki_test
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	certificatesv1 "k8s.io/api/certificates/v1"
+
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+)
+
+func TestGenerateTemplateFromCertificateSigningRequest(t *testing.T) {
+	csr, pk, err := gen.CSR(x509.RSA, gen.SetCSRDNSNames("example.com", "foo.example.com"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := map[string]struct {
+		csr            *certificatesv1.CertificateSigningRequest
+		expCertificate *x509.Certificate
+		expErr         bool
+	}{
+		"a CSR that contains an invalid duration should return an error": {
+			csr: gen.CertificateSigningRequest("",
+				gen.SetCertificateSigningRequestDuration("bad-duration"),
+				gen.SetCertificateSigningRequestUsages([]certificatesv1.KeyUsage{
+					certificatesv1.UsageKeyEncipherment,
+					certificatesv1.UsageDigitalSignature,
+				}),
+				gen.SetCertificateSigningRequestRequest(csr),
+			),
+			expCertificate: nil,
+			expErr:         true,
+		},
+		"a CSR that contains invalid usages should return an error": {
+			csr: gen.CertificateSigningRequest("",
+				gen.SetCertificateSigningRequestDuration("bad-duration"),
```

Comment:
suggestion: should this duration be valid?

Reply:
Yep, thanks.